### PR TITLE
[Chat] [Feature]  Clear Typing indicator

### PR DESCRIPTION
**Pull request by v-loalbert**
## Purpose

clear Typing indicator on inactivity timeout

clear Typing indicator when typing participant's message is received

clear Typing indicator when participant removed from chat

support having same display name for multiple participants

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Pull Request Type
What kind of change does this Pull Request introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Pull Request Checklist

- [ ] Public API changes
- [ ] Verified for cross-platform
- [ ] Synced with iOS, Web team
- [ ] Internal review completed
- [ ] UI Changes
  - [ ] Screen captures included
  - [ ] Tested for Light/Dark mode
  - [ ] Tested for screen rotation
  - [ ] Tested on Tablet and small screen device (5")
  - [ ] Include localization changes
- [ ] Tests
  - [ ] Memory leak analysis performed
  - [ ] Tested on API 21, 26 and latest 
  - [ ] Tested for background mode
  - [ ] Unit Tests Included
  - [ ] UI Automated Tests Included

## How to Test

* Open...
* Click on...

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

**Comment by iaulakh** [comment]
@v-loalbert can we use https://github.com/microsoft/fluentui-android/blob/master/fluentui_others/src/main/java/com/microsoft/fluentui?

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt [hunk 1]
```diff
@@ -9,4 +9,26 @@ import org.threeten.bp.OffsetDateTime
 internal data class ParticipantTimestampInfoModel(
     val userIdentifier: CommunicationIdentifier,
     val receivedOn: OffsetDateTime,
-) : BaseInfoModel
+) : BaseInfoModel {
+    companion object {
+        fun fromMessageInfoModel(infoModel: MessageInfoModel): ParticipantTimestampInfoModel =
```
Comment (iaulakh) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt:14]: would it be possible to move this logic outside?

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/DispatchTypingIndicatorReceived.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package com.azure.android.communication.ui.chat.redux.middleware.sdk
```
Comment (iaulakh) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/DispatchTypingIndicatorReceived.kt:1]: missing copyright

**Review by iaulakh — CHANGES_REQUESTED** [review]

**Comment by v-loalbert** [comment]
> @v-loalbert can we use https://github.com/microsoft/fluentui-android/blob/master/fluentui_others/src/main/java/com/microsoft/fluentui?

@iaulakh  Do you mean the Fluent UI's AvatarGroupView?
If so, then yes, I will use that instead of AvatarView to display the Typing indicator's Avatars

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt (hunk 1 above)
Comment (v-loalbert) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt:14]: @iaulakh what destination do you suggest for that block of code

Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt:14]: We can create helper class Albert.

Comment (v-loalbert) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/models/ParticipantTimestampInfoModel.kt]: done

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/DispatchTypingIndicatorReceived.kt (hunk 2 above)
Comment (v-loalbert) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/DispatchTypingIndicatorReceived.kt:4]: added

**Review by iaulakh — CHANGES_REQUESTED** [review]
Web shows only 1 is typing. Mobile shows 4
... is missing
space is not hidden when typing is collapsed
on message send participant is not removed
Screen Shot 2022-10-25 at 10 36 31 AM

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt [hunk 3]
```diff
@@ -87,7 +88,17 @@ internal class ChatServiceListener(
             is MessageInfoModel -> {
                 when (it.eventType) {
                     ChatEventType.CHAT_MESSAGE_RECEIVED -> {
-                        dispatch(ChatAction.MessageReceived(message = it.infoModel))
+                        val infoModel = it.infoModel
+                        dispatch(ChatAction.MessageReceived(message = infoModel))
+                        it.infoModel.senderCommunicationIdentifier?.let {
+                            dispatch(
+                                ParticipantAction.RemoveParticipantTyping(
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:95]: Parameter should probably just be user id

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt [hunk 4]
```diff
@@ -101,10 +112,17 @@ internal class ChatServiceListener(
             is ParticipantTimestampInfoModel -> {
                 when (it.eventType) {
                     ChatEventType.TYPING_INDICATOR_RECEIVED -> {
-                        dispatch(ParticipantAction.TypingIndicatorReceived(message = it.infoModel))
+                        val infoModel = it.infoModel
+                        dispatch(ParticipantAction.AddParticipantTyping(infoModel = infoModel))
+                        coroutineScope.launch {
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:117]: do we need logic here to cancel this scope if a new Typing indicator comes in for the same participant?

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt [hunk 5]
```diff
@@ -101,10 +112,17 @@ internal class ChatServiceListener(
             is ParticipantTimestampInfoModel -> {
                 when (it.eventType) {
                     ChatEventType.TYPING_INDICATOR_RECEIVED -> {
-                        dispatch(ParticipantAction.TypingIndicatorReceived(message = it.infoModel))
+                        val infoModel = it.infoModel
+                        dispatch(ParticipantAction.AddParticipantTyping(infoModel = infoModel))
```
Comment (ahammer) [inline azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:116]: Same here, AddParticipantTyping probably should just take a UID of user.

**Review by Mohtasim — DISMISSED** [review]

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt (hunk 4 above)
Comment (ahammer) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:117]: Also, Magic number, break out to a constant or val.

**Review by iaulakh — DISMISSED** [review]

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt (hunk 5 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:117]: We need both id+timestamp to identify unique notification

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt (hunk 4 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:118]: We can do in future to improve performance. The notification will do nothing as participant does not exist

File: azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt (hunk 3 above)
Comment (iaulakh) [reply azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/redux/middleware/sdk/ChatServiceListener.kt:96]: We need both id+timestamp to identify unique notification

**Review by ShaunaSong — DISMISSED** [review]

**Review by iaulakh — APPROVED** [review]

**Review by ahammer — APPROVED** [review]

**Pull request merged**